[Navigation] Improve maintaining the entries list and their interaction with HistoryItems

TingPing opened the pull request:
#### 7e54fb8ea939d9f0dc5d697d9daf410b7cb3b47f
<pre>
[Navigation] Improve maintaining the entries list and their interaction with HistoryItems
<a href="https://bugs.webkit.org/show_bug.cgi?id=273956">https://bugs.webkit.org/show_bug.cgi?id=273956</a>

Reviewed by Alex Christensen.

This changes a few areas but it all adds up to solving the same problems.
I&apos;ll go over the changes by entry point:

- navigation.navigate() takes a historyHandling argument that controls the
  history behavior all the way down.
- navigation.back/forward/traverseTo load the known HistoryItem instead of
  loading a URL as well as specify the correct FrameLoadType.
- the location APIs use the proper historyHandling.
- navigation.entries() is more accurately maintained:
  - On reactivation (loading a cached page from history) we regenerate the entires.
  - On a cross-document navigation we filter out any cross-origin entries.
- history.pushState/history.replaceState changes are reflected in entries.
  - This also no longer changes the key of replaced entries.
- Traversal events use the proper HistoryItem.

This all leads to a more consistent and correct list of entries. In order for adding
these entries to work there were also a few dispatches of NavigateEvent in the
expected places like during a history traversal or history state change.

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/currententrychange-event/history-back-same-doc-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/currententrychange-event/history-pushState-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/currententrychange-event/history-replaceState-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/currententrychange-event/navigate-from-initial-about-blank-same-doc-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/currententrychange-event/navigate-from-initial-about-blank-same-doc-popup-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/currententrychange-event/navigation-navigate-replace-same-doc-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigate-event/cross-origin-traversal-does-not-fire-navigate-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigate-event/cross-window/click-samedocument-crossorigin-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigate-event/cross-window/click-samedocument-sameorigin-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigate-event/event-constructor-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigate-event/intercept-and-navigate-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigate-event/intercept-history-pushState-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigate-event/intercept-history-replaceState-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigate-event/intercept-on-synthetic-event-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigate-event/intercept-popstate-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigate-event/intercept-same-document-history-back-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigate-event/navigate-destination-after-detach-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigate-event/navigate-destination-dynamic-index-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigate-event/navigate-destination-getState-back-forward-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigate-event/navigate-history-back-after-fragment-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigate-event/navigate-history-back-after-pushState-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigate-event/navigate-history-back-noop-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigate-event/navigate-history-pushState-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigate-event/navigate-history-replaceState-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigate-event/navigate-navigation-back-same-document-in-iframe-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigate-event/navigate-window-open-self-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigate-event/navigation-back-cross-document-preventDefault-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigate-event/navigation-traverseTo-in-iframe-same-document-preventDefault-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigate-event/replaceState-inside-back-handler-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-activation/activation-same-document-then-cross-document-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-history-entry/after-detach-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-history-entry/current-basic-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-history-entry/entries-after-bfcache-in-iframe-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-history-entry/index-not-in-entries-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-history-entry/key-id-back-same-document-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-history-entry/key-id-location-reload-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-history-entry/key-id-location-replace-cross-origin-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-history-entry/key-id-location-replace-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-history-entry/no-referrer-from-meta-url-censored-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-history-entry/no-referrer-url-censored-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-history-entry/sameDocument-after-fragment-navigate-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-history-entry/sameDocument-after-navigate-restore-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-history-entry/state-after-navigate-restore-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-methods/return-value/back-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-methods/return-value/back-intercept-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-methods/return-value/back-intercept-rejected-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-methods/return-value/forward-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-methods/return-value/forward-intercept-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-methods/return-value/forward-intercept-rejected-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-methods/return-value/navigate-intercept-interrupted-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-methods/return-value/navigate-interrupted-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-methods/return-value/navigate-unload-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-methods/return-value/reload-unload-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-methods/return-value/traverseTo-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-methods/return-value/traverseTo-intercept-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-methods/return-value/traverseTo-intercept-rejected-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-methods/traverseTo-after-adding-iframe-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-methods/traverseTo-after-data-url-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-methods/traverseTo-detach-between-navigate-and-navigatesuccess-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-methods/traverseTo-with-cross-origin-in-history-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/ordering-and-transition/currententrychange-dispose-ordering-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/per-entry-events/dispose-after-bfcache-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/per-entry-events/dispose-for-navigation-in-child-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/per-entry-events/dispose-same-document-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/per-entry-events/dispose-same-document-intercept-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/per-entry-events/dispose-same-document-navigate-during-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/per-entry-events/dispose-same-document-replaceState-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/per-entry-events/dispose-skip-current-on-truncate-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/state/cross-document-away-and-back-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/state/cross-document-location-api-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/state/location-reload-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/updateCurrentEntry-method/history-pushState-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/updateCurrentEntry-method/same-document-away-and-back-location-api-expected.txt:
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/navigation-api/navigate-event/cross-window/click-samedocument-crossorigin-sameorigindomain.sub-expected.txt:
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/navigation-api/navigation-history-entry/entries-after-navigations-in-multiple-windows-expected.txt:
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/navigation-api/navigation-methods/sandboxing-navigate-parent-expected.txt:
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::navigateFromServiceWorker):
* Source/WebCore/history/BackForwardController.cpp:
(WebCore::BackForwardController::allItems):
* Source/WebCore/history/BackForwardController.h:
* Source/WebCore/history/CachedPage.cpp:
(WebCore::CachedPage::restore):
* Source/WebCore/history/HistoryItem.cpp:
(WebCore::HistoryItem::HistoryItem):
* Source/WebCore/history/HistoryItem.h:
(WebCore::HistoryItem::uuidIdentifier const):
(WebCore::HistoryItem::operator== const):
* Source/WebCore/loader/FrameLoadRequest.h:
(WebCore::FrameLoadRequest::navigationHistoryBehavior const):
(WebCore::FrameLoadRequest::setNavigationHistoryHandling):
* Source/WebCore/loader/FrameLoader.cpp:
(WebCore::FrameLoader::determineNavigationType):
(WebCore::FrameLoader::loadInSameDocument):
(WebCore::FrameLoader::loadURL):
(WebCore::FrameLoader::loadWithDocumentLoader):
(WebCore::FrameLoader::continueFragmentScrollAfterNavigationPolicy):
(WebCore::FrameLoader::loadItem):
(WebCore::FrameLoader::updateNavigationAPIEntries):
(WebCore::determineNavigationType): Deleted.
* Source/WebCore/loader/FrameLoader.h:
* Source/WebCore/loader/HistoryController.cpp:
(WebCore::HistoryController::pushState):
(WebCore::HistoryController::replaceState):
* Source/WebCore/loader/NavigationScheduler.cpp:
(WebCore::NavigationScheduler::scheduleLocationChange):
* Source/WebCore/loader/NavigationScheduler.h:
* Source/WebCore/loader/SubframeLoader.cpp:
(WebCore::FrameLoader::SubframeLoader::loadOrRedirectSubframe):
* Source/WebCore/page/DOMWindow.h:
* Source/WebCore/page/History.cpp:
(WebCore::History::stateObjectAdded):
* Source/WebCore/page/LocalDOMWindow.cpp:
(WebCore::LocalDOMWindow::setLocation):
* Source/WebCore/page/LocalDOMWindow.h:
* Source/WebCore/page/Location.cpp:
(WebCore::Location::replace):
(WebCore::Location::setLocation):
* Source/WebCore/page/Navigation.cpp:
(WebCore::getEntryIndexOfHistoryItem):
(WebCore::Navigation::initializeEntries):
(WebCore::Navigation::maybeSetUpcomingNonTraversalTracker):
(WebCore::Navigation::navigate):
(WebCore::Navigation::performTraversal):
(WebCore::Navigation::traverseTo):
(WebCore::Navigation::back):
(WebCore::Navigation::forward):
(WebCore::Navigation::updateForNavigation):
(WebCore::Navigation::updateForReactivation):
(WebCore::Navigation::promoteUpcomingAPIMethodTracker):
(WebCore::Navigation::cleanupAPIMethodTracker):
(WebCore::Navigation::innerDispatchNavigateEvent):
(WebCore::Navigation::dispatchTraversalNavigateEvent):
* Source/WebCore/page/Navigation.h:
* Source/WebCore/page/NavigationHistoryEntry.cpp:
(WebCore::NavigationHistoryEntry::NavigationHistoryEntry):
(WebCore::NavigationHistoryEntry::sameDocument const):
(WebCore::NavigationHistoryEntry::getState const):
(WebCore::NavigationHistoryEntry::setState):
* Source/WebCore/page/NavigationHistoryEntry.h:
* Source/WebCore/page/RemoteDOMWindow.cpp:
(WebCore::RemoteDOMWindow::setLocation):
* Source/WebCore/page/RemoteDOMWindow.h:

Canonical link: <a href="https://commits.webkit.org/278960@main">https://commits.webkit.org/278960@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1a835172948293de10108b2c842d5f46309214db

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/52090 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/31418 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/4506 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/55363 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/2802 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/54394 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/37807 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/2510 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/42388 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/1786 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/54184 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/29047 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/44939 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/23459 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/26311 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/2216 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/972 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/48218 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/2361 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/56960 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/27209 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/2378 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/49785 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/28445 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/45060 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/49009 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/11399 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/29350 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/28186 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

charliewolfe commented on Source/WebCore/history/HistoryItem.h:
```diff
@@ -283,6 +287,7 @@ class HistoryItem : public RefCounted<HistoryItem>, public CanMakeWeakPtr<Histor
 #endif
 
     BackForwardItemIdentifier m_identifier;
+    WTF::UUID m_uuidIdentifier;
```
Why do you need to move this to `HistoryItem`? It looks like the only place this is accessed is from `NavigationHistoryEntry`. Can't we just keep it there?

TingPing replied:
The point of the key is that it refers to a specific HistoryItem, not a specific NavigationHistoryEntry which could be re-created.

https://html.spec.whatwg.org/multipage/nav-history-apis.html#the-navigationhistoryentry-interface

charliewolfe commented on Source/WebCore/page/Navigation.cpp:
```diff
@@ -85,12 +87,32 @@ bool Navigation::canGoForward() const
 }
 
+// https://html.spec.whatwg.org/multipage/nav-history-apis.html#getting-the-navigation-api-entry-index
+static std::optional<size_t> getEntryIndexOfHistoryItem(Vector<Ref<NavigationHistoryEntry>>& entries, const Ref<HistoryItem>& item)
```
We don't usually pass smart pointers as a const ref. Change to rvalue or remove Ref<>.

charliewolfe commented on Source/WebCore/page/Navigation.cpp:
```diff
@@ -85,12 +87,32 @@ bool Navigation::canGoForward() const
+}
+
 // https://html.spec.whatwg.org/multipage/nav-history-apis.html#initialize-the-navigation-api-entries-for-a-new-document
 void Navigation::initializeEntries(const Ref<HistoryItem>& currentItem, Vector<Ref<HistoryItem>>& items)
```
```suggestion
void Navigation::initializeEntries(Ref<HistoryItem>&& currentItem, Vector<Ref<HistoryItem>>& items)
```

charliewolfe commented on Source/WebCore/page/Navigation.cpp:
```diff
@@ -85,12 +87,32 @@ bool Navigation::canGoForward() const
         m_entries.append(NavigationHistoryEntry::create(protectedScriptExecutionContext().get(), item));
-    m_currentEntryIndex = items.find(currentItem);
+
+    m_currentEntryIndex = getEntryIndexOfHistoryItem(m_entries, currentItem);
```
```suggestion
    m_currentEntryIndex = getEntryIndexOfHistoryItem(m_entries, WTFMove(currentItem));
```

charliewolfe commented on Source/WebCore/page/Navigation.cpp:
```diff
@@ -437,6 +455,40 @@ void Navigation::updateForNavigation(Ref<HistoryItem>&& item, NavigationNavigati
 }
 
+// https://html.spec.whatwg.org/multipage/nav-history-apis.html#update-the-navigation-api-entries-for-reactivation
+void Navigation::updateForReactivation(Vector<Ref<HistoryItem>>& newHistoryItems, Ref<HistoryItem>& reactivatedItem)
```
```suggestion
void Navigation::updateForReactivation(Vector<Ref<HistoryItem>>& newHistoryItems, Ref<HistoryItem>&& reactivatedItem)
```

charliewolfe commented on Source/WebCore/page/Navigation.cpp:
```diff
@@ -437,6 +455,40 @@ void Navigation::updateForNavigation(Ref<HistoryItem>&& item, NavigationNavigati
+    }
+
+    m_entries = WTFMove(newEntries);
+    m_currentEntryIndex = getEntryIndexOfHistoryItem(m_entries, reactivatedItem);
```
```suggestion
    m_currentEntryIndex = getEntryIndexOfHistoryItem(m_entries, WTFMove(reactivatedItem));
```

charliewolfe commented on Source/WebCore/page/Navigation.h:
```diff
@@ -135,6 +132,7 @@ class Navigation final : public RefCounted<Navigation>, public EventTarget, publ
     bool dispatchDownloadNavigateEvent(const URL&, const String& downloadFilename);
 
     void updateForNavigation(Ref<HistoryItem>&&, NavigationNavigationType);
+    void updateForReactivation(Vector<Ref<HistoryItem>>& newHistoryItems, Ref<HistoryItem>& reactivatedItem);
```
```suggestion
    void updateForReactivation(Vector<Ref<HistoryItem>>& newHistoryItems, Ref<HistoryItem>&& reactivatedItem);
```

charliewolfe commented on Source/WebCore/page/Navigation.h:
```diff
@@ -26,6 +26,7 @@
 #pragma once
 
 #include "EventTarget.h"
+#include "FrameLoader.h"
```
I don't think you need this include.

charliewolfe commented on Source/WebCore/loader/FrameLoader.cpp:
```diff
@@ -4735,17 +4749,41 @@ void FrameLoader::updateNavigationAPIEntries()
+
+    auto rawEntries = page->backForward().allItems();
+    ssize_t startingIndex = rawEntries.find(*currentItem);
+    if (startingIndex != -1) {
```
```suggestion
    auto startingIndex = rawEntries.find(*currentItem);
    if (startingIndex != notFound) {
```

charliewolfe commented on Source/WebCore/loader/FrameLoadRequest.h:
```diff
@@ -111,6 +112,8 @@ class FrameLoadRequest {
     OptionSet<AdvancedPrivacyProtections> advancedPrivacyProtections() const { return m_advancedPrivacyProtections; }
 
+    NavigationHistoryBehavior navigationHistoryBehavior() const { return m_navigationHistoryBehavior; }
+    void setNavigationHistoryHandling(NavigationHistoryBehavior historyHandling) { m_navigationHistoryBehavior = historyHandling; }
```
Can the getter and setter for this variable have the same name? Instead of `HistoryHandling` and `HistoryBehavior`.

charliewolfe commented on Source/WebCore/history/CachedPage.cpp:
```diff
@@ -180,6 +182,13 @@ void CachedPage::restore(Page& page)
     }
 
+    if (localMainFrame && localMainFrame->settings().navigationAPIEnabled()) {
+        CheckedRef backForwardController = page.checkedBackForward();
```
```suggestion
        CheckedRef backForwardController = page.backForward();
```

charliewolfe commented on Source/WebCore/history/CachedPage.cpp:
```diff
@@ -180,6 +182,13 @@ void CachedPage::restore(Page& page)
+        CheckedRef backForwardController = page.checkedBackForward();
+        Ref currentItem = *backForwardController->currentItem();
+        auto allItems = backForwardController->allItems();
+        focusedDocument->domWindow()->navigation().updateForReactivation(allItems, currentItem);
```
`domWindow()` may need a null check.

charliewolfe commented on Source/WebCore/loader/FrameLoader.h:
```diff
@@ -93,6 +93,8 @@ enum class UsedLegacyTLS : bool;
 enum class IsMainResource : bool { No, Yes };
 enum class ShouldUpdateAppInitiatedValue : bool { No, Yes };
+enum class NavigationHistoryBehavior : uint8_t { Auto, Push, Replace };
+enum class NavigationNavigationType : uint8_t;
```
This forward declaration doesn't look like it's needed.

charliewolfe commented on Source/WebCore/loader/NavigationScheduler.h:
```diff
@@ -30,6 +30,7 @@
 
 #pragma once
 
+#include "FrameLoader.h"
```
Can you forward declare `NavigationHistoryBehavior` instead of including this?

charliewolfe commented on Source/WebCore/page/DOMWindow.h:
```diff
@@ -26,6 +26,7 @@
 #pragma once
 
 #include "EventTarget.h"
+#include "FrameLoader.h"
```
Remove this include and forward declare `NavigationHistoryBehavior`

charliewolfe commented on Source/WebCore/loader/FrameLoader.cpp:
```diff
@@ -1283,6 +1286,9 @@ void FrameLoader::loadInSameDocument(URL url, RefPtr<SerializedScriptValue> stat
     m_frame->checkedHistory()->updateForSameDocumentNavigation();
 
+    if (document->settings().navigationAPIEnabled())
+        document->protectedWindow()->navigation().updateForNavigation(*m_frame->history().currentItem(), navigationType);
```
Is it guaranteed that `protectedWindow()` and `currentItem()` are not null?

achristensen07 commented on Source/WebCore/history/CachedPage.cpp:
```diff
@@ -180,6 +182,13 @@ void CachedPage::restore(Page& page)
             frameView->updateContentsSize();
     }
 
+    if (localMainFrame && localMainFrame->settings().navigationAPIEnabled() && focusedDocument->domWindow()) {
```
Getting to the settings through the page would work better with site isolation, where the main frame might not be a LocalFrame.

achristensen07 commented on Source/WebCore/history/HistoryItem.cpp:
```diff
@@ -55,6 +55,7 @@ HistoryItem::HistoryItem(Client& client, const String& urlString, std::optional<
     , m_originalURLString(urlString)
     , m_pruningReason(PruningReason::None)
     , m_identifier(identifier ? *identifier : BackForwardItemIdentifier::generate())
+    , m_uuidIdentifier(WTF::UUID::createVersion4Weak())
```
Why are we using a UUID instead of an ObjectIdentifier or ObjectIdentifier or ProcessQualified<ObjectIdentifier>?  Why are we using weak random values?

TingPing replied:
The spec requires the format of an entry's key to be a UUID.

I used weak because it doesn't really have a security concern.

achristensen07 commented on Source/WebCore/loader/FrameLoadRequest.h:
```diff
@@ -133,6 +135,7 @@ class FrameLoadRequest {
     bool m_isRequestFromClientOrUserInput { false };
     bool m_isInitialFrameSrcLoad { false };
     OptionSet<AdvancedPrivacyProtections> m_advancedPrivacyProtections;
+    NavigationHistoryBehavior m_navigationHistoryBehavior;
```
Let's give this a default value so we don't have uninitialized memory.

achristensen07 commented on Source/WebCore/loader/FrameLoader.cpp:
```diff
@@ -1578,8 +1574,8 @@ void FrameLoader::loadURL(FrameLoadRequest&& frameLoadRequest, const String& ref
         RELEASE_ASSERT(!isBackForwardLoadType(newLoadType) || frame->history().provisionalItem());
-        policyChecker().checkNavigationPolicy(WTFMove(request), ResourceResponse { } /* redirectResponse */, oldDocumentLoader.get(), WTFMove(formState), [this, frame, requesterOrigin = Ref { frameLoadRequest.requesterSecurityOrigin() }] (const ResourceRequest& request, WeakPtr<FormState>&&, NavigationPolicyDecision navigationPolicyDecision) {
-            continueFragmentScrollAfterNavigationPolicy(request, requesterOrigin.ptr(), navigationPolicyDecision == NavigationPolicyDecision::ContinueLoad);
+        policyChecker().checkNavigationPolicy(WTFMove(request), ResourceResponse { } /* redirectResponse */, oldDocumentLoader.get(), WTFMove(formState), [this, frame, requesterOrigin = Ref { frameLoadRequest.requesterSecurityOrigin() }, historyHandling = frameLoadRequest.navigationHistoryBehavior()] (const ResourceRequest& request, WeakPtr<FormState>&&, NavigationPolicyDecision navigationPolicyD ...
```
Could we add a CheckedPtr to this or a Ref<Frame> to keep this alive in a separate PR?  I don't see a reason why this would not be destroyed by the time the lambda is called and uses this.  Separate PR because it isn't related to this PR and we may want to revert it separately if it causes problems.

achristensen07 commented on Source/WebCore/loader/FrameLoader.cpp:
```diff
@@ -4311,7 +4307,15 @@ void FrameLoader::loadDifferentDocumentItem(HistoryItem& item, HistoryItem* from
-    RefPtr currentItem = m_frame->history().currentItem();
+    RefPtr currentItem = frame().history().currentItem();
+
+    if (frame().document()->settings().navigationAPIEnabled() && fromItem && SecurityOrigin::create(item.url())->isSameOriginAs(SecurityOrigin::create(fromItem->url()))) {
```
May as well null check document.

achristensen07 commented on Source/WebCore/loader/FrameLoader.cpp:
```diff
@@ -4742,17 +4746,40 @@ void FrameLoader::updateNavigationAPIEntries()
+                break;
+
+            entriesForNavigationAPI.insert(0, WTFMove(item));
+            i -= 1;
```
This is what -- is for.  Also, stylistically this would be a little nicer as a for loop.  The variable i doesn't even need to be the same variable as the next loop because its value is reset.

achristensen07 commented on Source/WebCore/loader/FrameLoader.cpp:
```diff
@@ -4742,17 +4746,40 @@ void FrameLoader::updateNavigationAPIEntries()
+
+        int64_t i = startingIndex - 1;
+        while (i >= 0) {
+            Ref item = rawEntries[i];
```
I feel like this needs more bounds checks.  startingIndex can be 0, then i would be -1, and I'm not sure if we're doing the right thing here.

TingPing replied:
I believe its correct but I agree this can be cleaned up. The style was basically word for word the spec.

(if the item is the first in the list, `(startingIndex - 1 >= 0)` will just be false and no items are inserted before it)

achristensen07 commented on Source/WebCore/loader/FrameLoader.cpp:
```diff
@@ -4742,17 +4746,40 @@ void FrameLoader::updateNavigationAPIEntries()
+        }
+
+        i = startingIndex + 1;
+        while (static_cast<size_t>(i) < rawEntries.size()) {
```
Also would be better as a for loop.

achristensen07 commented on Source/WebCore/loader/HistoryController.cpp:
```diff
@@ -929,6 +930,9 @@ void HistoryController::pushState(RefPtr<SerializedScriptValue>&& stateObject, c
     addVisitedLink(*page, URL({ }, urlString));
     frame->checkedLoader()->client().updateGlobalHistory();
+
+    if (frame->document()->settings().navigationAPIEnabled())
```
Null check document.

achristensen07 commented on Source/WebCore/loader/HistoryController.cpp:
```diff
@@ -956,6 +960,9 @@ void HistoryController::replaceState(RefPtr<SerializedScriptValue>&& stateObject
     addVisitedLink(*page, URL({ }, urlString));
     frame->checkedLoader()->client().updateGlobalHistory();
+
+    if (frame->document()->settings().navigationAPIEnabled())
```
Null check document.

achristensen07 commented on Source/WebCore/page/History.cpp:
```diff
@@ -272,6 +273,12 @@ ExceptionOr<void> History::stateObjectAdded(RefPtr<SerializedScriptValue>&& data
         return Exception { ExceptionCode::SecurityError, makeString("Attempt to use history.pushState() more than "_s, perStateObjectTimeSpanLimit, " times per "_s, stateObjectTimeSpan.seconds(), " seconds"_s) };
     }
 
+    if (frame->document()->settings().navigationAPIEnabled()) {
```
Null check document.

achristensen07 commented on Source/WebCore/page/Location.cpp:
```diff
@@ -305,8 +305,13 @@ ExceptionOr<void> Location::setLocation(LocalDOMWindow& incumbentWindow, LocalDO
 
+    // https://html.spec.whatwg.org/multipage/nav-history-apis.html#the-location-interface:location-object-navigate
+    auto historyHandling = NavigationHistoryBehavior::Auto;
+    if (!firstFrame->loader().isComplete() && !firstFrame->document()->domWindow()->hasTransientActivation())
```
Null checks.

achristensen07 commented on Source/WebCore/page/Navigation.cpp:
```diff
@@ -290,7 +306,7 @@ Navigation::Result Navigation::performTraversal(const String& key, Navigation::O
     auto entry = findEntryByKey(key);
     ASSERT(entry);
-    frame()->loader().load(FrameLoadRequest(*frame(), URL(entry.value()->url())));
+    frame()->loader().loadItem(entry.value()->associatedHistoryItem(), &currentEntry()->associatedHistoryItem(), loadType, ShouldTreatAsContinuingLoad::No);
```
Should we be using currentEntry and entry here?  Can they be different.

TingPing replied:
First value is `target`, second value is `from` (the current).

The case of being the same is handled above, so they are always different.

achristensen07 commented on Source/WebCore/page/Navigation.cpp:
```diff
@@ -442,6 +460,40 @@ void Navigation::updateForNavigation(Ref<HistoryItem>&& item, NavigationNavigati
+        return;
+
+    Vector<Ref<NavigationHistoryEntry>> newEntries;
+    Vector<Ref<NavigationHistoryEntry>> oldEntries = WTFMove(m_entries);
```
std::exchange leaves m_entries in a defined state.

achristensen07 commented on Source/WebCore/page/Navigation.cpp:
```diff
@@ -442,6 +460,40 @@ void Navigation::updateForNavigation(Ref<HistoryItem>&& item, NavigationNavigati
+        if (!newEntry)
+            newEntry = NavigationHistoryEntry::create(scriptExecutionContext(), WTFMove(item));
+
+        newEntries.append(newEntry.releaseNonNull());
```
Do we want to append to the end if we removed from the middle?

TingPing replied:
This is in a loop creating a new list so it adds everything that had a shared history linearly.

The final step below sets the active entry (`m_currentEntryIndex = getEntryIndexOfHistoryItem(m_entries, reactivatedItem);`

achristensen07 commented on Source/WebCore/history/BackForwardController.cpp:
```diff
@@ -170,6 +170,17 @@ RefPtr<HistoryItem> BackForwardController::itemAtIndex(int i)
+{
+    Vector<Ref<HistoryItem>> historyItems;
+    for (int index = -1 * static_cast<int>(backCount()); index <= static_cast<int>(forwardCount()); index++) {
+        if (RefPtr item = itemAtIndex(index))
```
else ASSERT_NOT_REACHED()

TingPing replied:
I've hit this condition before.

My assumption was that `WebBackForwardListProxy` is to another process so there is a possible race condition between getting the size and an item.

TingPing replied:
I guess the better solution here is a new API that returns a whole list from the client. I think a race still exists, but the items are more likely to be accurate. I can do a follow-up if that makes sense.

achristensen07 commented on Source/WebCore/loader/HistoryController.cpp:
```diff
@@ -956,6 +960,9 @@ void HistoryController::replaceState(RefPtr<SerializedScriptValue>&& stateObject
     addVisitedLink(*page, URL({ }, urlString));
     frame->checkedLoader()->client().updateGlobalHistory();
+
+    if (auto* document = frame->document(); document->settings().navigationAPIEnabled())
```
RefPtr instead of auto*

achristensen07 commented on Source/WebCore/page/History.cpp:
```diff
@@ -272,6 +273,12 @@ ExceptionOr<void> History::stateObjectAdded(RefPtr<SerializedScriptValue>&& data
         return Exception { ExceptionCode::SecurityError, makeString("Attempt to use history.pushState() more than "_s, perStateObjectTimeSpanLimit, " times per "_s, stateObjectTimeSpan.seconds(), " seconds"_s) };
     }
 
+    if (auto* document = frame->document(); document->settings().navigationAPIEnabled()) {
```
RefPtr instead of auto*
null check document before calling Document::settings

achristensen07 commented on Source/WebCore/page/Navigation.cpp:
```diff
@@ -83,12 +85,32 @@ bool Navigation::canGoForward() const
 }
 
+// https://html.spec.whatwg.org/multipage/nav-history-apis.html#getting-the-navigation-api-entry-index
+static std::optional<size_t> getEntryIndexOfHistoryItem(Vector<Ref<NavigationHistoryEntry>>& entries, const HistoryItem& item)
```
const Vector

achristensen07 commented on Source/WebCore/page/Navigation.cpp:
```diff
@@ -83,12 +85,32 @@ bool Navigation::canGoForward() const
+static std::optional<size_t> getEntryIndexOfHistoryItem(Vector<Ref<NavigationHistoryEntry>>& entries, const HistoryItem& item)
+{
+    size_t index = 0;
+    for (auto& entry : entries) {
```
Ideally we would have a better algorithm eventually rather than just searching through a Vector.

TingPing committed:
[Navigation] Improve maintaining the entries list and their interaction with HistoryItems
https://bugs.webkit.org/show_bug.cgi?id=273956

Reviewed by Alex Christensen.

This changes a few areas but it all adds up to solving the same problems.
I'll go over the changes by entry point:

- navigation.navigate() takes a historyHandling argument that controls the
  history behavior all the way down.
- navigation.back/forward/traverseTo load the known HistoryItem instead of
  loading a URL as well as specify the correct FrameLoadType.
- the location APIs use the proper historyHandling.
- navigation.entries() is more accurately maintained:
  - On reactivation (loading a cached page from history) we regenerate the entires.
  - On a cross-document navigation we filter out any cross-origin entries.
- history.pushState/history.replaceState changes are reflected in entries.
  - This also no longer changes the key of replaced entries.
- Traversal events use the proper HistoryItem.

This all leads to a more consistent and correct list of entries. In order for adding
these entries to work there were also a few dispatches of NavigateEvent in the
expected places like during a history traversal or history state change.

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/currententrychange-event/history-back-same-doc-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/currententrychange-event/history-pushState-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/currententrychange-event/history-replaceState-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/currententrychange-event/navigate-from-initial-about-blank-same-doc-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/currententrychange-event/navigate-from-initial-about-blank-same-doc-popup-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/currententrychange-event/navigation-navigate-replace-same-doc-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigate-event/cross-origin-traversal-does-not-fire-navigate-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigate-event/cross-window/click-samedocument-crossorigin-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigate-event/cross-window/click-samedocument-sameorigin-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigate-event/event-constructor-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigate-event/intercept-and-navigate-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigate-event/intercept-history-pushState-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigate-event/intercept-history-replaceState-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigate-event/intercept-on-synthetic-event-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigate-event/intercept-popstate-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigate-event/intercept-same-document-history-back-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigate-event/navigate-destination-after-detach-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigate-event/navigate-destination-dynamic-index-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigate-event/navigate-destination-getState-back-forward-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigate-event/navigate-history-back-after-fragment-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigate-event/navigate-history-back-after-pushState-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigate-event/navigate-history-back-noop-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigate-event/navigate-history-pushState-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigate-event/navigate-history-replaceState-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigate-event/navigate-navigation-back-same-document-in-iframe-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigate-event/navigate-window-open-self-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigate-event/navigation-back-cross-document-preventDefault-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigate-event/navigation-traverseTo-in-iframe-same-document-preventDefault-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigate-event/replaceState-inside-back-handler-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-activation/activation-same-document-then-cross-document-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-history-entry/after-detach-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-history-entry/current-basic-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-history-entry/entries-after-bfcache-in-iframe-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-history-entry/index-not-in-entries-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-history-entry/key-id-back-same-document-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-history-entry/key-id-location-reload-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-history-entry/key-id-location-replace-cross-origin-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-history-entry/key-id-location-replace-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-history-entry/no-referrer-from-meta-url-censored-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-history-entry/no-referrer-url-censored-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-history-entry/sameDocument-after-fragment-navigate-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-history-entry/sameDocument-after-navigate-restore-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-history-entry/state-after-navigate-restore-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-methods/return-value/back-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-methods/return-value/back-intercept-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-methods/return-value/back-intercept-rejected-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-methods/return-value/forward-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-methods/return-value/forward-intercept-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-methods/return-value/forward-intercept-rejected-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-methods/return-value/navigate-intercept-interrupted-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-methods/return-value/navigate-interrupted-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-methods/return-value/navigate-unload-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-methods/return-value/reload-unload-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-methods/return-value/traverseTo-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-methods/return-value/traverseTo-intercept-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-methods/return-value/traverseTo-intercept-rejected-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-methods/traverseTo-after-adding-iframe-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-methods/traverseTo-after-data-url-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-methods/traverseTo-detach-between-navigate-and-navigatesuccess-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-methods/traverseTo-with-cross-origin-in-history-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/ordering-and-transition/currententrychange-dispose-ordering-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/per-entry-events/dispose-after-bfcache-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/per-entry-events/dispose-for-navigation-in-child-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/per-entry-events/dispose-same-document-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/per-entry-events/dispose-same-document-intercept-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/per-entry-events/dispose-same-document-navigate-during-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/per-entry-events/dispose-same-document-replaceState-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/per-entry-events/dispose-skip-current-on-truncate-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/state/cross-document-away-and-back-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/state/cross-document-location-api-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/state/location-reload-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/updateCurrentEntry-method/history-pushState-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/updateCurrentEntry-method/same-document-away-and-back-location-api-expected.txt:
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/navigation-api/navigate-event/cross-window/click-samedocument-crossorigin-sameorigindomain.sub-expected.txt:
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/navigation-api/navigation-history-entry/entries-after-navigations-in-multiple-windows-expected.txt:
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/navigation-api/navigation-methods/sandboxing-navigate-parent-expected.txt:
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::navigateFromServiceWorker):
* Source/WebCore/history/BackForwardController.cpp:
(WebCore::BackForwardController::allItems):
* Source/WebCore/history/BackForwardController.h:
* Source/WebCore/history/CachedPage.cpp:
(WebCore::CachedPage::restore):
* Source/WebCore/history/HistoryItem.cpp:
(WebCore::HistoryItem::HistoryItem):
* Source/WebCore/history/HistoryItem.h:
(WebCore::HistoryItem::uuidIdentifier const):
(WebCore::HistoryItem::operator== const):
* Source/WebCore/loader/FrameLoadRequest.h:
(WebCore::FrameLoadRequest::navigationHistoryBehavior const):
(WebCore::FrameLoadRequest::setNavigationHistoryHandling):
* Source/WebCore/loader/FrameLoader.cpp:
(WebCore::FrameLoader::determineNavigationType):
(WebCore::FrameLoader::loadInSameDocument):
(WebCore::FrameLoader::loadURL):
(WebCore::FrameLoader::loadWithDocumentLoader):
(WebCore::FrameLoader::continueFragmentScrollAfterNavigationPolicy):
(WebCore::FrameLoader::loadItem):
(WebCore::FrameLoader::updateNavigationAPIEntries):
(WebCore::determineNavigationType): Deleted.
* Source/WebCore/loader/FrameLoader.h:
* Source/WebCore/loader/HistoryController.cpp:
(WebCore::HistoryController::pushState):
(WebCore::HistoryController::replaceState):
* Source/WebCore/loader/NavigationScheduler.cpp:
(WebCore::NavigationScheduler::scheduleLocationChange):
* Source/WebCore/loader/NavigationScheduler.h:
* Source/WebCore/loader/SubframeLoader.cpp:
(WebCore::FrameLoader::SubframeLoader::loadOrRedirectSubframe):
* Source/WebCore/page/DOMWindow.h:
* Source/WebCore/page/History.cpp:
(WebCore::History::stateObjectAdded):
* Source/WebCore/page/LocalDOMWindow.cpp:
(WebCore::LocalDOMWindow::setLocation):
* Source/WebCore/page/LocalDOMWindow.h:
* Source/WebCore/page/Location.cpp:
(WebCore::Location::replace):
(WebCore::Location::setLocation):
* Source/WebCore/page/Navigation.cpp:
(WebCore::getEntryIndexOfHistoryItem):
(WebCore::Navigation::initializeEntries):
(WebCore::Navigation::maybeSetUpcomingNonTraversalTracker):
(WebCore::Navigation::navigate):
(WebCore::Navigation::performTraversal):
(WebCore::Navigation::traverseTo):
(WebCore::Navigation::back):
(WebCore::Navigation::forward):
(WebCore::Navigation::updateForNavigation):
(WebCore::Navigation::updateForReactivation):
(WebCore::Navigation::promoteUpcomingAPIMethodTracker):
(WebCore::Navigation::cleanupAPIMethodTracker):
(WebCore::Navigation::innerDispatchNavigateEvent):
(WebCore::Navigation::dispatchTraversalNavigateEvent):
* Source/WebCore/page/Navigation.h:
* Source/WebCore/page/NavigationHistoryEntry.cpp:
(WebCore::NavigationHistoryEntry::NavigationHistoryEntry):
(WebCore::NavigationHistoryEntry::sameDocument const):
(WebCore::NavigationHistoryEntry::getState const):
(WebCore::NavigationHistoryEntry::setState):
* Source/WebCore/page/NavigationHistoryEntry.h:
* Source/WebCore/page/RemoteDOMWindow.cpp:
(WebCore::RemoteDOMWindow::setLocation):
* Source/WebCore/page/RemoteDOMWindow.h:

Canonical link: https://commits.webkit.org/278960@main
